experimental/sql/dialects/rel_ssa.py

```diff
@@ -0,0 +1,271 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from typing import Any, Union, List
+from xdsl.ir import Block, Region, Operation, SSAValue, ParametrizedAttribute, Data, MLContext, Attribute
+from xdsl.dialects.builtin import StringAttr, ArrayAttr, ArrayOfConstraint, IntegerAttr
+from xdsl.irdl import AttributeDef, OperandDef, ResultDef, RegionDef, SingleBlockRegionDef, irdl_attr_definition, irdl_op_definition, ParameterDef, AnyAttr, VarOperandDef, builder
+
+# This file contains the relational ssa dialect, a dialect that expresses
+# relational queries using SSA. The dialect consists of two types of operations:
+# Expressions and Operators. Expressions work on single values, whereas
+# Operators work on Bags. The Column and Yield Expressions bridge the gap to
+# convert from one to the other.
+
+
+@irdl_attr_definition
+class DataType(ParametrizedAttribute):
+  """
+  Models a datatype in a relational ssa query.
+  """
+  name = "rel_ssa.datatype"
+
+
+@irdl_attr_definition
+class Int32(DataType):
+  """
+  Models a int32 type in a relational ssa query.
+
+  Example:
+
+  ```
+  !rel_ssa.int32
+  ```
+  """
+  name = "rel_ssa.int32"
+
+
+@irdl_attr_definition
+class String(DataType):
+  """
+  Models a string type in a relational ssa query, that can be either nullable or
+  not.
+
+  Example:
+
+  ```
+  !rel_ssa.string<0: !i1>
+  ```
+  """
+  name = "rel_ssa.string"
+
+  nullable = ParameterDef(IntegerAttr)
+
+  @staticmethod
+  @builder
+  def get(val: Union[int, IntegerAttr]) -> 'String':
+    if isinstance(val, IntegerAttr):
+      return String([val])
+    return String([IntegerAttr.from_int_and_width(val, 1)])
+
+
+@irdl_attr_definition
+class Boolean(ParametrizedAttribute):
+  """
+  Models a type that can either be true or false to, e.g., show whether a tuple
+  is part of the result or not.
+
+  Example:
+
+  '''
+  !rel_ssa.bool
+  '''
+  """
+  name = "rel_ssa.bool"
+
+
+@irdl_attr_definition
+class Bag(ParametrizedAttribute):
```

Comment:
In the meeting, we discussed having named tuples as part of this type. I have yet to investigate how this will work using xDSL.

Reply:
OK, cool! If it's not trivial, feel free to leave it as is in this PR and change later.

Reply:
I went with an approach that introduces an attribute `SchemaElement` that contains a `StringAttr` for the name and a `DataType` for the type. This makes the IR quite cluttered, but I feel we are only custom printer away from what we actually would want (i.e. syntax of the form `*name* : *type*`).